python_modules/dagster/dagster/_config/structured_config.py

```diff
@@ -113,12 +123,90 @@ def _curry_config_schema(schema_field: Field, data: Any) -> IDefinitionConfigSch
 
 
 ResValue = TypeVar("ResValue")
+IOManagerValue = TypeVar("IOManagerValue", bound=IOManager)
+
+import pydantic
+
+
+# Since a metaclass is invoked by Resource before Resource or PartialResource is defined, we need to
+# define a temporary class to use as a placeholder for use in the initial metaclass invocation.
+# When the metaclass is invoked for a Resource subclass, it will use the non-placeholder values.
+class _Temp(Generic[ResValue]):
+    pass
+
+
+_ResourceDep: Type = _Temp
+_Resource: Type = _Temp
+_PartialResource: Type = _Temp
+
+
+@dataclass_transform()
+class BaseResourceMeta(pydantic.main.ModelMetaclass):
+    """
+    Custom metaclass for Resource and PartialResource. This metaclass is responsible for
+    transforming the type annotations on the class to allow for partially configured resources.
+    """
+
+    def __new__(self, name, bases, namespaces, **kwargs):
+        # Gather all type annotations from the class and its base classes
+        annotations = namespaces.get("__annotations__", {})
+        for base in bases:
+            if hasattr(base, "__annotations__"):
+                annotations.update(base.__annotations__)
+        for field in annotations:
+            if not field.startswith("__"):
+                # Check if the annotation is a ResourceDependency
+                if get_origin(annotations[field]) == _ResourceDep:
+                    # arg = get_args(annotations[field])[0]
+                    # If so, we treat it as a Union of a PartialResource and a Resource
+                    # for Pydantic's sake.
+                    annotations[field] = Any
+                elif _safe_is_subclass(annotations[field], _Resource):
+                    # If the annotation is a Resource, we treat it as a Union of a PartialResource
+                    # and a Resource for Pydantic's sake, so that a user can pass in a partially
+                    # configured resource.
+                    base = annotations[field]
+                    annotations[field] = Union[_PartialResource[base], base]
+
+        namespaces["__annotations__"] = annotations
+        return super().__new__(self, name, bases, namespaces, **kwargs)
```

Comment:
We should have a separate file for all the code that effects only the typehinting this the Description mixin and this metaclass. The entire file will do nothing other than the annotation mangling. Again I'm also envisioning a future where rip out this stuff and want to be able to do it cleanly.

Reply:
Moved metaclass to a standalone file